### PR TITLE
Add Replan operation to restart changed services.

### PR DESCRIPTION
**Pull request by hpidcock**
This change adds the **replan** command for restarting services that have changed or removed (we don't have remove AFAIK yet) since when they were started.

Replan acts much the same as **autostart** with the added restarting of changes.

Based on #58

Compare changes:
https://github.com/canonical/pebble/compare/hpidcock:restart...hpidcock:replan

### CODE REVIEW

File: internal/overlord/servstate/handlers.go [hunk 1]
```diff
@@ -77,17 +78,37 @@ func (m *ServiceManager) doStart(task *state.Task, tomb *tomb.Tomb) error {
...
+	}
+
+	// Pass service description's environment variables to child process
+	env := os.Environ()
+	for k, v := range service.Environment {
+		env = append(env, k+"="+v)
+	}
+
+	// Expand the command string with environment variables.
+	envMap := make(map[string]string)
```
Comment (niemeyer) [inline internal/overlord/servstate/handlers.go:92]: This looks unrelated to this PR. I'm keen on supporting this ASAP as it introduces an incompatibility, but this is something worth talking about to make sure we get it right.

For example, how should "mycmd "$ASD"` expand?

I suggest taking that out of the PR so we can merge this, and let's discuss this by itself.

File: internal/overlord/servstate/handlers.go [hunk 2]
```diff
@@ -126,14 +141,22 @@ func (m *ServiceManager) doStart(task *state.Task, tomb *tomb.Tomb) error {
...
 	go func() {
 		active.err = cmd.Wait()
 		_ = active.logBuffer.Close()
+		releasePlan, err := m.acquirePlan()
+		if err == nil {
+			if m.services[req.Name].cmd == cmd {
+				delete(m.services, req.Name)
+			}
+			releasePlan()
+		}
```
Comment (niemeyer) [inline internal/overlord/servstate/handlers.go:159]: But that change means we are now losing the metadata about the command termination. With the previous logic, we can comfortable look at the metadata for the running state at any point, while with this change it will be gone as soon as the service itself dies, so we in fact can never trust the cmd data to be there anymore because we don't control the process termination. Am I missing something?  If not and there wasn't a clear motivation for this change, I suggest reverting it for the time being.

File: internal/overlord/servstate/manager.go [hunk 3]
```diff
@@ -327,3 +342,52 @@ func (m *ServiceManager) ServiceLogs(services []string, last int) (map[string]se
...
+	releasePlan, err := m.acquirePlan()
+	if err != nil {
+		return nil, nil, err
+	}
+	defer releasePlan()
+
+	needsRestart := make(map[string]bool)
+	stopOrder := []string(nil)
+	for name, service := range m.services {
+		if currentPlan, ok := m.plan.Services[name]; ok &&
```
Comment (niemeyer) [inline internal/overlord/servstate/manager.go:358]: currentService?

**Review by niemeyer — CHANGES_REQUESTED** [review]
Looks very nice, thanks!

Some comments, but nothing fundamental.

File: client/services.go [hunk 4]
```diff
@@ -41,6 +41,16 @@ func (client *Client) Stop(opts *ServiceOptions) (changeID string, err error) {
 	return changeID, err
 }
 
+func (client *Client) Restart(opts *ServiceOptions) (changeID string, err error) {
+	_, changeID, err = client.doMultiServiceAction("restart", opts.Names)
+	return changeID, err
+}
+
+func (client *Client) Replan(opts *ServiceOptions) (changeID string, err error) {
+	_, changeID, err = client.doMultiServiceAction("replan", opts.Names)
```
[the remark this answers is not in the document]
Comment (benhoyt) [reply client/services.go:50]: I definitely prefer "replan" (though it was weird to me at first!). "Comply" sounds a bit authoritarian :-), plus comply is vague (comply with what?) whereas "replan" links it so nicely to "plan".

File: internal/overlord/servstate/manager.go [hunk 5]
```diff
@@ -12,7 +12,21 @@ import (
 	"github.com/canonical/pebble/internal/servicelog"
 )
 
-type ServiceManager struct {
+type ServiceManager interface {
```
Comment (benhoyt) [inline internal/overlord/servstate/manager.go:15]: Curious why you changed this to an interface? Normally in Go packages return concrete types, and interfaces are only created where needed (e.g., in `Overlord` itself, so the `ServiceManager` can be mocked out).

File: internal/overlord/servstate/manager.go [hunk 6]
```diff
@@ -327,3 +342,52 @@ func (m *ServiceManager) ServiceLogs(services []string, last int) (map[string]se
...
+// their plans had changed between when they started and this call.
+func (m *serviceManager) Replan() ([]string, []string, error) {
+	releasePlan, err := m.acquirePlan()
+	if err != nil {
+		return nil, nil, err
+	}
+	defer releasePlan()
+
+	needsRestart := make(map[string]bool)
+	stopOrder := []string(nil)
```
Comment (benhoyt) [inline internal/overlord/servstate/manager.go:356]: Nit: in my experience `var stopOrder []string` is more common for the zero (nil) value. Not sure exactly why, but it is (maybe because there's less punctuation?). Same below.

For example, in Juju there are 1371 instances of `var foo []T`, but only 3 of `foo := []T(nil)`. Similar in the Go stdlib.

File: internal/plan/plan.go [hunk 7]
```diff
@@ -57,6 +58,38 @@ type Service struct {
...
+	}
+	return &copy
+}
+
+// Equal returns true when the two services are equal in value.
+func (s *Service) Equal(other *Service) bool {
+	if s == other {
+		return true
+	}
+	// TODO: implement this efficiently.
```
Comment (benhoyt) [inline internal/plan/plan.go:89]: Probably not a concern for this use case?

File: cmd/pebble/cmd_replan.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"github.com/canonical/pebble/client"
+	"github.com/jessevdk/go-flags"
+)
+
+var shortReplanHelp = "Restart running services that have changed and start default services"
```
[the remark this answers is not in the document]
Comment (hpidcock) [reply cmd/pebble/cmd_replan.go:22]: Much better description thank-you

File: internal/overlord/servstate/handlers.go (hunk 1 above)
Comment (hpidcock) [reply internal/overlord/servstate/handlers.go:92]: This is in #58 so I'll remove it from there and propose a separate PR we can drill into this more with.

File: cmd/pebble/cmd_restart.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+import (
+	"github.com/jessevdk/go-flags"
+
+	"github.com/canonical/pebble/client"
+)
+
+var shortRestartHelp = "Restart a service and its dependencies"
+var longRestartHelp = `
+The restart command restarts the service with the provided name and
+any other services it depends on, in the correct order.
```
[the remark this answers is not in the document]
Comment (hpidcock) [reply cmd/pebble/cmd_restart.go:26]: StopOrder adds in dependents. So therefore the stop command stops dependents.
I think we should be consistent here, should dependents always stop/restart for the replan/stop/restart commands?

I don't mind either way, just is strange behavior. Starts make sense to happen in order with dependencies, to reduce failure on the critical path. Dependents should be fault tolerant and handle failure gracefully, if not the health-check work and auto-restarts should handle that.

File: internal/overlord/servstate/manager.go [hunk 10]
```diff
@@ -25,10 +25,11 @@ type ServiceManager struct {
 }
 
 type activeService struct {
-	cmd       *exec.Cmd
-	err       error
-	done      chan struct{}
-	logBuffer *servicelog.RingBuffer
+	plannedService *plan.Service
```
[the remark this answers is not in the document]
Comment (hpidcock) [reply internal/overlord/servstate/manager.go:28]: I've gone with originalService, hopefully that is sufficient.

File: cmd/pebble/cmd_restart.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (hpidcock) [reply cmd/pebble/cmd_restart.go]: StopOrder can optionally not add dependants to the returned slice. This fixes the strange behaviour in restart & replan

File: client/services.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (hpidcock) [reply client/services.go:50]: Replan is the cleanest option I think, happy to continue with it.

File: cmd/pebble/cmd_restart.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (niemeyer) [reply cmd/pebble/cmd_restart.go]: This is being sorted on the underlying #58.

File: internal/overlord/servstate/manager.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (niemeyer) [reply internal/overlord/servstate/manager.go]: Thanks :+1:

File: internal/plan/plan.go [hunk 11]
```diff
@@ -172,21 +205,21 @@ func CombineLayers(layers ...*Layer) (*Layer, error) {
...
+func (p *Plan) StopOrder(names []string, includeDependants bool) ([]string, error) {
+	return order(p.Services, names, true, includeDependants)
 }
 
-func order(services map[string]*Service, names []string, stop bool) ([]string, error) {
+func order(services map[string]*Service, names []string, stop bool, includeRelated bool) ([]string, error) {
 	// For stop, create a list of reversed dependencies.
 	predecessors := map[string][]string(nil)
-	if stop {
+	if stop && includeRelated {
```
Comment (niemeyer) [inline internal/plan/plan.go:222]: While reviewing the follow up PR, I started to suspect that this logic can't be right, as it's blocking ordering knowledge from the algorithm altogether.  We need to feed the hierarchy into the dependency algorithm even if we're not going to be using all of it, because we don't know which paths are present in the graph. In other words, even if we only care about A and D, we need to have A, B, C, and D into the dependency logic, so that we can find the path betweem A and D and order them.  As such, I suggest leaving StartOrder and StopOrder alone, and just having a helper function that when given two lists L1 and L2, will just reorder L1 according to the order of these elements L2.

**Review by niemeyer — CHANGES_REQUESTED** [review]
Oops, sorry, one additional comment that probably needs looking into:

File: internal/overlord/servstate/handlers.go [hunk 12]
```diff
@@ -75,9 +75,14 @@ func (m *ServiceManager) doStart(task *state.Task, tomb *tomb.Tomb) error {
 		return fmt.Errorf("cannot find service %q in plan", req.Name)
 	}
 
-	_, previous := m.services[req.Name]
-	if previous {
-		return fmt.Errorf("service %q was previously started", req.Name)
+	if service, previous := m.services[req.Name]; previous {
+		select {
+		case <-service.done:
+			// Service exited but not yet removed, safe to override.
```
Comment (niemeyer) [inline internal/overlord/servstate/handlers.go:81]: This is only safe because we check for the cmd field before actually removing. We might probably give a hint here so future generations don't fall into the trap of thinking that this is always safe, and we can also be more explicit about what it is that it's safe to do. Something along the lines of:
```
// Service exited but not yet removed. It's safe to replace it
// because the removal will check if the cmd still matches.
```

File: internal/plan/plan.go [hunk 13]
```diff
@@ -172,21 +205,21 @@ func CombineLayers(layers ...*Layer) (*Layer, error) {
...
+	return order(p.Services, names, false, true)
 }
 
 // StopOrder returns the required services that must be stopped for the named
 // services to be properly stopped, in the order that they must be stopped.
 // An error is returned when a provided service name does not exist, or there
 // is an order cycle involving the provided service or its dependencies.
-func (p *Plan) StopOrder(names []string) ([]string, error) {
-	return order(p.Services, names, true)
+func (p *Plan) StopOrder(names []string, includeDependants bool) ([]string, error) {
```
Comment (niemeyer) [inline internal/plan/plan.go:215]: This needs to be looked into per note on #58: I suggest leaving the original logic in the package alone, and just having that sorting function that applies sorting from a list into a different subset list.

**Review by niemeyer — CHANGES_REQUESTED** [review]
Only blocker here is the underlying #58.O

Otherwise, a trivial and LGTM, thanks!

File: internal/daemon/api_services.go [hunk 14]
```diff
@@ -82,33 +87,89 @@ func v1PostServices(c *Command, r *http.Request, _ *userState) Response {
...
+			replanned[v] = true
+		}
+		for _, v := range startNames {
+			replanned[v] = true
+		}
+		for k := range replanned {
+			services = append(services, k)
+		}
+		sort.Strings(services)
+		payload.Services = services
```
Comment (niemeyer) [inline internal/daemon/api_services.go:172]: This reads very cleanly, thank you.

**Review by niemeyer — APPROVED** [review]
LGTM, thanks!

**Comment by jnsgruk** [comment]
Nice work! Can we just make sure we get this covered in the docs over at https://juju.is/docs/sdk/pebble please? :)
